### PR TITLE
Refactored GoGeneAnnotationParser, modified and extended tests

### PR DESCRIPTION
**Pull request by pnrobinson**
Hi Jules! I am hoping to working through a lot of the outstanding issues in the next few weeks and hopefully to make a release. This PR addresses: https://github.com/monarch-initiative/phenol/issues/240

### CODE REVIEW

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java [hunk 1]
```diff
@@ -35,104 +37,57 @@
...
+      do {
+        final String[] arry = nextLine.split("\t");
+        try {
+          GoGaf21Annotation annot = new GoGaf21Annotation(arry);
+          builder.add(annot);
+        } catch (PhenolException e) {
+          System.err.printf("[ERROR] Could not parse GoGaf line (%s): %s", nextLine, e.getMessage());
+          // just skip this error. Should actually never happen
+        }
+      } while ((nextLine = br.readLine()) != null);
```
Comment (julesjacobsen) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: you're calling `nextLine = br.readLine()` a three times in the `try` block. Why not just call it once? 

```java
try (BufferedReader br = new  BufferedReader(new FileReader(file))) {
    while ((nextLine = br.readLine()) != null) {
        // skip the comments, which start with exclamation marks
        if (nextLine.isEmpty() || nextLine.startsWith("!")) {
            continue;         
        } else {
            final String[] arry = nextLine.split("\t");
            try {
                GoGaf21Annotation annot = new GoGaf21Annotation(arry);
                builder.add(annot);
...

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java [hunk 2]
```diff
@@ -35,104 +37,57 @@
...
+    String nextLine;
+    ImmutableList.Builder<GoGaf21Annotation> builder = new ImmutableList.Builder<>();
+    try (BufferedReader br = new  BufferedReader(new FileReader(file))) {
+      nextLine =br.readLine();
+      // skip the comments, which start with exclamation marks
+      while (nextLine.isEmpty() || nextLine.startsWith("!")) {
+        nextLine = br.readLine();
+      }
+      do {
+        final String[] arry = nextLine.split("\t");
```
Comment (julesjacobsen) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: Do you need to check the array length?

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java [hunk 3]
```diff
@@ -35,104 +37,57 @@
...
-   *
-   */
-
-  private void skipUntilData() throws IOException {
-    while (nextLine == null || nextLine.startsWith("!")) {
-      nextLine = reader.readLine();
+  public static List<GoGaf21Annotation> loadAnnotations(File file) {
+    String nextLine;
+    ImmutableList.Builder<GoGaf21Annotation> builder = new ImmutableList.Builder<>();
+    try (BufferedReader br = new  BufferedReader(new FileReader(file))) {
```
Comment (julesjacobsen) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: could replace 

```java
BufferedReader br = new  BufferedReader(new FileReader(file))
```
with 

```java
BufferedReader br = Files.newBufferedReader(path)
```

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java [hunk 4]
```diff
@@ -35,104 +37,57 @@
...
 
-  /*
-   * Skip comment lines until data
-   *
-   */
-
-  private void skipUntilData() throws IOException {
-    while (nextLine == null || nextLine.startsWith("!")) {
-      nextLine = reader.readLine();
+  public static List<GoGaf21Annotation> loadAnnotations(File file) {
```
Comment (julesjacobsen) [inline phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: Replace `File` with `Path` unless you have a strong reason not to.

File: phenol-cli/src/main/java/org/monarchinitiative/phenol/cli/demo/GoEnrichmentDemo.java [hunk 5]
```diff
@@ -50,8 +50,8 @@ public void run() {
       int n_terms = gontology.countAllTerms();
       System.out.println("[INFO] parsed " + n_terms + " GO terms.");
       System.out.println("[INFO] parsing  " + pathGoGaf);
-      final GoGeneAnnotationParser annotparser = new GoGeneAnnotationParser(pathGoGaf);
-      List<TermAnnotation> goAnnots = annotparser.getTermAnnotations();
+     // final GoGeneAnnotationParser annotparser = new GoGeneAnnotationParser(pathGoGaf);
```
Comment (julesjacobsen) [inline phenol-cli/src/main/java/org/monarchinitiative/phenol/cli/demo/GoEnrichmentDemo.java]: remember to remove the commented out line

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java (hunk 1 above)
Comment (pnrobinson) [reply phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: Thanks for this tidbit of elegant code, Master Jules, much appreciated :-0 !!

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java (hunk 2 above)
Comment (pnrobinson) [reply phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: This gets checked in the constructor of the GaGofAnnotation. Perhaps it is better just to pass the line as a String and split it within the constructor? But it is checked.

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java (hunk 1 above)
Comment (julesjacobsen) [reply phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: (bow)

File: phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java (hunk 2 above)
Comment (julesjacobsen) [reply phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: Yeah, that would be better - you could write a static GaGofAnnotation.parseAnnotation(String line) method to do all that.

Comment (pnrobinson) [reply phenol-annotations/src/main/java/org/monarchinitiative/phenol/annotations/obo/go/GoGeneAnnotationParser.java]: I will refactor as suggested before merging, thanks!

**Comment by pnrobinson** [comment]
thanks, I have implemented the changes, tests are passed.

**Pull request merged**